src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -105,7 +114,7 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
                 _applicationLifetime.StopApplication();
 
                 IList<Exception> exceptions = new List<Exception>();
-                if (_hostedServices != null) // Started?
+                if (_hostedServices.Count > 0) // Started?
                 {
                     foreach (IHostedService hostedService in _hostedServices.Reverse())
```

Comment:
You've hit [List<T>.Reverse](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.reverse?view=net-5.0#System_Collections_Generic_List_1_Reverse), instead of the original LINQ call:
```suggestion
                    foreach (IHostedService hostedService in _hostedServices.AsEnumerable().Reverse())
```

Reply:
Thanks for solving that mystery 🙏 I believe this was introduced as I accepted [this suggestion](https://github.com/dotnet/runtime/pull/52517/commits/720f78362c8edc5b281fe2eddaeb2f1b0a645b48). Is there an argument for reverting that change instead of applying this fix? Changing back to use `ICollection<IHostedService>` will result in a smaller diff in the MR, but perhaps it is worth applying your fix as it as it is more explicit. Any thought on this?

Reply:
Change the loop to a for loop and enumerate in reverse

Reply:
That's a great suggestion 👌 I've added a new commit to the PR with this change. Let me know if you prefer fixups instead.